staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go

```diff
@@ -1426,35 +1434,120 @@ func (c *cacheWatcher) process(ctx context.Context, resourceVersion uint64) {
 	}
 }
 
+// ready is a condition variable that represent readiness and
+// allow callers to wait on its status, returning an error with the reason.
+// It allows to use context cancellation and to be stopped,
+// on both cases it stops waiting raising the corresponding error.
+
 type ready struct {
-	ok bool
-	c  *sync.Cond
+	mu      sync.Mutex
+	readyCh chan struct{}
+	stopCh  chan struct{}
 }
 
+// newReady creates a ready condition variable that is not stopped and not ready
 func newReady() *ready {
-	return &ready{c: sync.NewCond(&sync.RWMutex{})}
+	return &ready{
+		readyCh: make(chan struct{}), // is ready if it does not exist
+		stopCh:  make(chan struct{}), // is stopped if it does not exist
+	}
 }
 
-func (r *ready) wait() {
-	r.c.L.Lock()
-	for !r.ok {
-		r.c.Wait()
+// wait blocks until:
+// - it is ready and returns nil
+// - it is stopped and returns an error
+// - context is cancelled and returns error
+func (r *ready) wait(ctx context.Context) error {
+	select {
+	// if it is ready but context is cancelled
+	// context cancellation fails fast.
+	case <-ctx.Done():
+		return ctx.Err()
+	default:
+		// readyCh and stopCh must not be closed at the same time
```

Comment:
nit: can we avoid nesting here?

Reply:
is an edge case, but we do need the nesting or it flakes if it is ready and the context is cancelled, 50% returns the error of the context and the other 50% returns the ready